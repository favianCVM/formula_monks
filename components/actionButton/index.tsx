import React from 'react';
import {Pressable, PressableProps} from 'react-native';
import {Text} from '../fragments/Text';
import Icon from 'react-native-vector-icons/FontAwesome6';
import {COLORS} from '../../styles/colors';

interface ActionButtonProps extends Omit<PressableProps, 'children'> {
  icon: string;
  children: string;
}

const ActionButton = React.memo(
  ({icon, children, style, ...props}: ActionButtonProps) => {
    return (
      <Pressable
        style={({pressed}) => ({
          borderColor: COLORS.text,
          borderWidth: 1.5,
          flexDirection: 'row',
          alignItems: 'center',
          alignSelf: 'flex-start',
          padding: 10,
          borderRadius: 10,
          backgroundColor: pressed ? COLORS.blue : COLORS.gray,
          ...(typeof style === 'object' ? style : {}),
        })}
        {...props}>
        <Icon
          color={COLORS.text}
          style={{marginRight: 10}}
          name={icon}
          size={20}
        />
        <Text>{children}</Text>
      </Pressable>
    );
  },
);

export {ActionButton};
