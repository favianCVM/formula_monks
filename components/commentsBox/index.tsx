import React from 'react';
import {PostComment} from '../../views/PostDetails';
import {CommentCard} from './CommentCard';
import {FlatList, GestureHandlerRootView} from 'react-native-gesture-handler';
import {Dimensions} from 'react-native';
import {normalize} from '../../libs/normalizeSize';
import {Title} from '../fragments/Title';

interface CommentsBoxProps {
  comments: Array<PostComment>;
}

const CommentsBox = React.memo(({comments}: CommentsBoxProps) => {
  return (
    <GestureHandlerRootView style={{flex: 1}}>
      <FlatList
        contentContainerStyle={{
          paddingHorizontal: normalize(15),
        }}
        data={comments}
        keyExtractor={({id}) => `post-comment-${id}`}
        renderItem={({item}) => <CommentCard {...item} />}
      />
    </GestureHandlerRootView>
  );
});

export {CommentsBox};
