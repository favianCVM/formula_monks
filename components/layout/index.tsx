import React from 'react';
import {View, ViewProps} from 'react-native';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import Toast from 'react-native-toast-message';

const Layout = React.memo(({children}: ViewProps) => {
  return (
    <SafeAreaProvider>
      {children}
      <Toast />
    </SafeAreaProvider>
  );
});

export {Layout};
