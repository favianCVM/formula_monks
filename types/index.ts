export interface Post {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export interface Rating {
  id: Post['id'];
  value: number;
}

export interface PostComment {
  postId: number;
  id: number;
  name: string;
  email: string;
  body: string;
}

export interface AuthorInformation {
  id: number;
  name: string;
  username: string;
  email: string;
  address: {
    street: string;
    city: string;
  };
  phone: string;
  website: string;
  company: {
    name: string;
  };
}

export interface PostDetails {
  id: Post['id'];
  author: AuthorInformation;
  comments: Array<PostComment>;
}
