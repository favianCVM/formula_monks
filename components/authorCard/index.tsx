import React from 'react';
import {View} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome6';
import {Text} from '../fragments/Text';
import {COLORS} from '../../styles/colors';
import { AuthorInformation } from '../../types';


const AuthorCard = React.memo(
  ({
    address,
    company,
    email,
    name,
    phone,
    username,
    website,
    id,
  }: AuthorInformation) => {
    return (
      <View style={{backgroundColor: COLORS.gray, padding: 18}}>
        <View
          style={{
            flexDirection: 'row',
            alignItems: 'center',
            marginBottom: 18,
          }}>
          <Icon
            style={{
              backgroundColor: COLORS.green,
              padding: 20,
              borderRadius: 10,
            }}
            name="user"
            size={30}
          />
          <View
            style={{
              flexDirection: 'column',
              alignItems: 'flex-start',
              flex: 1,
              paddingLeft: 20,
            }}>
            <Text>username: {username}</Text>
            <Text>name: {name}</Text>
            <Text>user id: {id}</Text>
          </View>
        </View>
        <View
          style={{
            flexDirection: 'row',
            justifyContent: 'space-between',
            alignItems: 'flex-end',
          }}>
          <View>
            <View
              style={{
                flexDirection: 'row',
                alignItems: 'center',
              }}>
              <Icon
                color={COLORS.text}
                name="envelope"
                size={20}
                style={{marginRight: 16}}
              />
              <Text>{email}</Text>
            </View>
            <View
              style={{
                flexDirection: 'row',
                alignItems: 'center',
                marginVertical: 6,
              }}>
              <Icon
                color={COLORS.text}
                name="phone"
                size={20}
                style={{marginRight: 16}}
              />
              <Text>{phone}</Text>
            </View>
            <View style={{flexDirection: 'row', alignItems: 'center'}}>
              <Icon
                color={COLORS.text}
                name="briefcase"
                size={20}
                style={{marginRight: 16}}
              />
              <Text>{company.name}</Text>
            </View>
          </View>
          <View>
            <Icon
              color={COLORS.text}
              name="location-dot"
              size={20}
              style={{alignSelf: 'center'}}
            />
            <Text>{address.city}</Text>
            <Text>{address.street}</Text>
          </View>
        </View>
      </View>
    );
  },
);

export {AuthorCard};
