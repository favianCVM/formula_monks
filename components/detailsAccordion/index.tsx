import React, {JSX} from 'react';
//@ts-ignore
import {AccordionList} from 'accordion-collapse-react-native';
import Icon from 'react-native-vector-icons/FontAwesome6';
import {CommentsBox} from '../commentsBox';
import {AuthorCard} from '../authorCard';
import {PostDetails} from '../../types';
import {View} from 'react-native';
import {Title} from '../fragments/Title';
import {COLORS} from '../../styles/colors';
import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  headerContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    borderColor: COLORS.text,
    borderWidth: 1,
    paddingVertical: 12,
  },
  headerIcon: {
    marginRight: 20,
    padding: 8,
    backgroundColor: COLORS.yellow,
    borderRadius: 12,
  },
  headerTitle: {
    fontSize: 20,
  },
  bodyContainer: {paddingVertical: 12, backgroundColor: COLORS.gray},
});

const Header = React.memo(({title}: {title: string}) => {
  return (
    <View style={styles.headerContainer}>
      <Icon
        name={title === 'Author information' ? 'user' : 'message'}
        size={20}
        style={styles.headerIcon}
      />
      <Title style={styles.headerTitle}>{title}</Title>
    </View>
  );
});

const Body = React.memo(({body}: {body: JSX.Element}) => (
  <View style={styles.bodyContainer}>{body}</View>
));

const DetailsAccordion = React.memo(({author, comments}: PostDetails) => (
  <AccordionList
    list={[
      {
        id: 1,
        title: 'Author information',
        body: <AuthorCard {...author} />,
      },
      {
        id: 2,
        title: 'Comments',
        body: <CommentsBox comments={comments} />,
      },
    ]}
    header={({title}: {title: string}) => <Header title={title} />}
    body={({body}: {body: JSX.Element}) => <Body body={body} />}
  />
));

export {DetailsAccordion};
