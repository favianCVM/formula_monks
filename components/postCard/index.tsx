import React from 'react';
import {removePost} from '../../store/reducers/posts';
import {Pressable, StyleSheet, Text} from 'react-native';
import {COLORS} from '../../styles/colors';
import {useAppDispatch} from '../../hooks';
import {addFavorite, removeFavorite} from '../../store/reducers/posts';
import {normalize} from '../../libs/normalizeSize';
import Icon from 'react-native-vector-icons/FontAwesome6';
import { Post } from '../../types';

interface PostCardProps extends Post {
  handleDetailsRedirection: (
    postId: Post['id'],
    userId: Post['userId'],
  ) => void;
  isFavorite: boolean;
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: 6,
    paddingVertical: 5,
    marginBottom: 10,
    borderRadius: 8,
    borderColor: COLORS.text,
    borderWidth: 1,
  },
  pressedContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: COLORS.blue,
    paddingHorizontal: 6,
    paddingVertical: 5,
    marginBottom: 10,
    borderRadius: 8,
    borderColor: COLORS.text,
    borderWidth: 1,
  },
  title: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    fontSize: normalize(20),
    color: COLORS.text,
  },
});

const PostCard = React.memo(
  ({
    title,
    id,
    userId,
    handleDetailsRedirection,
    isFavorite,
  }: PostCardProps) => {
    const dispatch = useAppDispatch();

    const handleFavoriteToggle = () => {
      dispatch(isFavorite ? removeFavorite(id) : addFavorite(id));
    };

    const handleDeletePost = () => {
      if (isFavorite) dispatch(removeFavorite(id));
      dispatch(removePost(id));
    };

    return (
      <Pressable
        testID="PostCard"
        style={({pressed}) =>
          pressed ? styles.pressedContainer : styles.container
        }
        onPress={() => handleDetailsRedirection(id, userId)}>
        <Text numberOfLines={1} adjustsFontSizeToFit style={styles.title}>
          {title}
        </Text>
        <Icon.Button
          style={{
            paddingVertical: 4,
            paddingHorizontal: 0,
            marginLeft: 4,
            borderRadius: 5,
          }}
          onPress={handleDeletePost}
          testID="PostCardDeletePressable"
          size={35}
          name="trash-can"
          backgroundColor={'transparent'}
          underlayColor={COLORS.red}
          color={COLORS.text}
        />
        <Icon.Button
          style={{
            paddingVertical: 4,
            paddingHorizontal: 0,
            marginLeft: 4,
            borderRadius: 5,
          }}
          onPress={handleFavoriteToggle}
          testID="PostCardFavoritePressable"
          size={35}
          name="heart"
          backgroundColor={'transparent'}
          underlayColor={COLORS.red}
          color={isFavorite ? COLORS.red : COLORS.text}
        />
      </Pressable>
    );
  },
);

export {PostCard};
