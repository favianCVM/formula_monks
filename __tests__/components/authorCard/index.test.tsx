import React from 'react';
import {render, screen} from '@testing-library/react-native';
import {AuthorCard} from '../../../components/authorCard/';
import {AuthorInformation} from '../../../types';
import renderer from 'react-test-renderer';

describe('Author information card element', () => {
  const authorInformationPlaceholder: AuthorInformation = {
    id: 1,
    address: {
      city: 'city-placeholder',
      street: 'street-placeholder',
    },
    company: {
      name: '',
    },
    email: 'email-placeholder',
    name: 'name-placeholder',
    phone: 'phone-placeholder',
    username: 'username-placeholder',
    website: 'website-placeholder',
  };

  it('renders author information card with the given values', () => {
    const {getByTestId} = render(
      <AuthorCard {...authorInformationPlaceholder} />,
    );

    const authorCardElement = getByTestId('AuthorCardElement');
    expect(authorCardElement).not.toBeNull();

    expect(
      screen.getByText(`${authorInformationPlaceholder.id}`),
    ).toBeVisible();
    expect(
      screen.getByText(authorInformationPlaceholder.address.city),
    ).toBeVisible();
    expect(
      screen.getByText(authorInformationPlaceholder.address.street),
    ).toBeVisible();
    expect(
      screen.getByText(authorInformationPlaceholder.company.name),
    ).toBeVisible();
    expect(screen.getByText(authorInformationPlaceholder.email)).toBeVisible();
    expect(screen.getByText(authorInformationPlaceholder.name)).toBeVisible();
    expect(screen.getByText(authorInformationPlaceholder.phone)).toBeVisible();
    expect(
      screen.getByText(authorInformationPlaceholder.username),
    ).toBeVisible();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<AuthorCard {...authorInformationPlaceholder} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
