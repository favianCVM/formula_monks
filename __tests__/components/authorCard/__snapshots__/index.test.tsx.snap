// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author information card element renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#F0ECE4",
      "padding": 18,
    }
  }
  testID="AuthorCardElement"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 18,
      }
    }
  >
    <Icon
      name="user"
      size={30}
      style={
        {
          "backgroundColor": "#0F9D58",
          "borderRadius": 10,
          "padding": 20,
        }
      }
    />
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
          "flexDirection": "column",
          "paddingLeft": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          username:
        </Text>
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          username-placeholder
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          name:
        </Text>
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          name-placeholder
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          user id:
        </Text>
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          1
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#1B1C1E"
          name="envelope"
          size={20}
          style={
            {
              "marginRight": 16,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          email-placeholder
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginVertical": 6,
          }
        }
      >
        <Icon
          color="#1B1C1E"
          name="phone"
          size={20}
          style={
            {
              "marginRight": 16,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        >
          phone-placeholder
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          color="#1B1C1E"
          name="briefcase"
          size={20}
          style={
            {
              "marginRight": 16,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "#1B1C1E",
                "fontSize": 16,
              },
              undefined,
            ]
          }
          testID="TextFragment"
        />
      </View>
    </View>
    <View>
      <Icon
        color="#1B1C1E"
        name="location-dot"
        size={20}
        style={
          {
            "alignSelf": "center",
          }
        }
      />
      <Text
        style={
          [
            {
              "color": "#1B1C1E",
              "fontSize": 16,
            },
            undefined,
          ]
        }
        testID="TextFragment"
      >
        city-placeholder
      </Text>
      <Text
        style={
          [
            {
              "color": "#1B1C1E",
              "fontSize": 16,
            },
            undefined,
          ]
        }
        testID="TextFragment"
      >
        street-placeholder
      </Text>
    </View>
  </View>
</View>
`;
